x-pack/plugins/reporting/server/export_types/common/pdf/worker.ts

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { Ensure, SerializableRecord } from '@kbn/utility-types';
+
+import { parentPort, isMainThread, MessagePort } from 'worker_threads';
+import _ from 'lodash';
+import path from 'path';
+import Printer from 'pdfmake';
+
+import { getTemplate } from './get_template';
+import type { TemplateLayout } from './types';
+import { assetPath } from './constants';
+
+export type PdfWorkerData = Ensure<
+  {
+    layout: TemplateLayout;
+    title: string;
+    content: SerializableRecord[];
+
+    logo?: string;
+  },
+  SerializableRecord
+>;
+
+export interface GeneratePdfRequest {
+  port: MessagePort;
+  data: PdfWorkerData;
+}
+
+export type GeneratePdfResponse = SuccessResponse | ErrorResponse;
+
+export interface SuccessResponse {
+  error?: undefined;
+  data: {
+    buffer: Uint8Array;
+    metrics: {
+      pages: number;
+    };
+  };
+}
+
+export interface ErrorResponse {
+  error: string;
+  data: null;
+}
+
+if (!isMainThread) {
+  parentPort!.on('message', execute);
+}
+
+const getPageCount = (pdfDoc: PDFKit.PDFDocument): number => {
+  const pageRange = pdfDoc.bufferedPageRange();
+  if (!pageRange) {
+    return 0;
+  }
+  const { count, start } = pageRange;
+
+  return start + count;
+};
+
+async function execute({ data: { layout, logo, title, content }, port }: GeneratePdfRequest) {
+  try {
+    const tableBorderWidth = 1;
+
+    const fontPath = (filename: string) => path.resolve(assetPath, 'fonts', filename);
+
+    const fonts = {
+      Roboto: {
+        normal: fontPath('roboto/Roboto-Regular.ttf'),
+        bold: fontPath('roboto/Roboto-Medium.ttf'),
+        italics: fontPath('roboto/Roboto-Italic.ttf'),
+        bolditalics: fontPath('roboto/Roboto-Italic.ttf'),
+      },
+      'noto-cjk': {
+        // Roboto does not support CJK characters, so we'll fall back on this font if we detect them.
+        normal: fontPath('noto/NotoSansCJKtc-Regular.ttf'),
+        bold: fontPath('noto/NotoSansCJKtc-Medium.ttf'),
+        italics: fontPath('noto/NotoSansCJKtc-Regular.ttf'),
+        bolditalics: fontPath('noto/NotoSansCJKtc-Medium.ttf'),
+      },
+    };
+
+    const printer = new Printer(fonts);
+
+    const docDefinition = _.assign(getTemplate(layout, logo, title, tableBorderWidth, assetPath), {
```

Comment:
It seems `docDefinition` could come from the workerData, giving the worker thread a little less work to do. WDYT?

Reply:
In this case I moved this to the worker because `getTemplate` can return an object that contains a function (`getFooter` is a function that returns a function because its content can be dynamic due to page numbers being injected).

I hadn't thought of how we might refactor this to make it all serialized, but I also think it's OK for the worker to do more work`. The more we can offload from the main thread the more benefit we will get from using a worker thread.

Reply:
I had in mind that it might not just be less work for the worker thread to do, but could be a cleaner separation. This comment here speaks better to having a clean separation: https://github.com/elastic/kibana/pull/124787/files#r807894719 so I'll defer to that.